sympy/sets/sets.py

```diff
@@ -1202,28 +1202,32 @@ def reduce(args):
 
         # ===== Pair-wise Rules =====
         # Here we depend on rules built into the constituent sets
-        args = set(args)
+        args = FiniteSet(*args)
         new_args = True
         while(new_args):
             for s in args:
                 new_args = False
-                for t in args - set((s,)):
+                args_without_s = set(args) - set(FiniteSet(s))
+                args_without_s = FiniteSet(*[ar for ar in args_without_s])
+                for t in args_without_s:
                     new_set = s._union(t)
                     # This returns None if s does not know how to intersect
                     # with t. Returns the newly intersected set otherwise
                     if new_set is not None:
                         if not isinstance(new_set, set):
                             new_set = set((new_set, ))
-                        new_args = (args - set((s, t))).union(new_set)
+                        remv = set((s, t))
+                        new_args = (set(args) - remv).union(new_set)
                         break
                 if new_args:
+                    new_args = FiniteSet(*[ar for ar in new_args])
                     args = new_args
```

Comment:
There is [`break`](https://github.com/sympy/sympy/pull/11188/files#diff-7b3e9b0852f1b1d094f87233f56e201fR1225)  after `args = new_args` . So it comes out of the   [Line 1208 `for` loop](https://github.com/sympy/sympy/pull/11188/files#diff-7b3e9b0852f1b1d094f87233f56e201fR1208) and since `while(new_args)`  is not none,   `for` loop [L1208] will start again with modified `args`.  

Whenever `new_set = s._union(t)` new_set is not none (means reduced form comes) , it will modify the `new_args` and `break` the loop, to start again fresh .

`while(new_args)` --  new_args is for just ([ `false`](https://github.com/sympy/sympy/pull/11188/files#diff-7b3e9b0852f1b1d094f87233f56e201fR1209) or not empty `FiniteSet`)  to check whether `args`  is modified or not.(means something reduced or not). If `args` is not modified(`new_args`  remain `false`) then `while` loop Line 1207 ends.

The issue will come when we do something like this : 

```
        while(new_args):
            for s in args:
                new_args = False
                args_without_s = set(args) - set(FiniteSet(s))
                args_without_s = FiniteSet(*[ar for ar in args_without_s])
                for t in args_without_s:
                    new_set = s._union(t)
                    # This returns None if s does not know how to intersect
                    # with t. Returns the newly intersected set otherwise
                    if new_set is not None:
                        if not isinstance(new_set, set):
                            new_set = set((new_set, ))
                        remv = set((s, t))
                        new_args = (set(args) - remv).union(new_set)
                        # this new_args should be used next time. `new_args` length is `len(args) - 1`. so this loop break.
                        # and goes to next if condition `if new_args:`.
                        break
                if new_args:
                    new_args = FiniteSet(*[ar for ar in new_args])
                    args = new_args
                    # if here is no break here, it will continue with the loop `for s in args` in Line 1208
                    # with this modified args. Which will return wrong result.

```